File: frontend/js/websocket.js
/**
 * WebSocket Client for Real-Time Dashboard Updates
 *
 * Handles:
 * - Connection management with auto-reconnect
 * - Real-time event broadcasting
 * - Toast notifications for updates
 */

class WebSocketClient {
    constructor() {
        this.ws = null;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 10;
        this.reconnectInterval = 5000; // 5 seconds
        this.heartbeatInterval = null;
        this.listeners = {};
        this.connectionStatus = 'disconnected';
    }

    /**
     * Connect to WebSocket server
     */
    connect() {
        const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
        const host = window.location.host;
        const wsUrl = `${protocol}//${host}/ws`;

        console.log(`Connecting to WebSocket: ${wsUrl}`);

        try {
            this.ws = new WebSocket(wsUrl);

            this.ws.onopen = () => {
                console.log('WebSocket connected');
                this.connectionStatus = 'connected';
                this.reconnectAttempts = 0;
                this.updateConnectionIndicator(true);
                this.startHeartbeat();
                this.emit('connected');
            };

            this.ws.onmessage = (event) => {
                try {
                    const message = JSON.parse(event.data);
                    this.handleMessage(message);
                } catch (error) {
                    console.error('Failed to parse WebSocket message:', error);
                }
            };

            this.ws.onclose = () => {
                console.log('WebSocket disconnected');
                this.connectionStatus = 'disconnected';
                this.updateConnectionIndicator(false);
                this.stopHeartbeat();
                this.emit('disconnected');
                this.attemptReconnect();
            };

            this.ws.onerror = (error) => {
                console.error('WebSocket error:', error);
                this.emit('error', error);
            };

        } catch (error) {
            console.error('Failed to create WebSocket connection:', error);
            this.attemptReconnect();
        }
    }

    /**
     * Disconnect from WebSocket
     */
    disconnect() {
        if (this.ws) {
            this.ws.close();
            this.ws = null;
        }
        this.stopHeartbeat();
    }

    /**
     * Attempt to reconnect with exponential backoff
     */
    attemptReconnect() {
        if (this.reconnectAttempts >= this.maxReconnectAttempts) {
            console.error('Max reconnect attempts reached');
            this.emit('max_reconnect_attempts');
            return;
        }

        this.reconnectAttempts++;
        const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 30000);

        console.log(`Reconnecting in ${delay/1000}s (attempt ${this.reconnectAttempts}/${this.maxReconnectAttempts})`);

        setTimeout(() => {
            this.connect();
        }, delay);
    }

    /**
     * Start heartbeat to keep connection alive
     */
    startHeartbeat() {
        this.heartbeatInterval = setInterval(() => {
            if (this.ws && this.ws.readyState === WebSocket.OPEN) {
                this.ws.send('ping');
            }
        }, 30000); // Every 30 seconds
    }

    /**
     * Stop heartbeat
     */
    stopHeartbeat() {
        if (this.heartbeatInterval) {
            clearInterval(this.heartbeatInterval);
            this.heartbeatInterval = null;
        }
    }

    /**
     * Handle incoming messages
     */
    handleMessage(message) {
        const { type, data, timestamp } = message;

        console.log('WebSocket message:', type, data);

        // Emit event to listeners
        this.emit(type, data, timestamp);

        // Show toast notification
        switch (type) {
            case 'new_analysis':
                this.showToast(
                    `New ${data.source} analysis`,
                    `${data.content_type} - ${data.sentiment} (${data.conviction}/10)`
                );
                break;

            case 'collection_complete':
                this.showToast(
                    `${data.source} collection complete`,
                    `Collected ${data.items_collected} items`
                );
                break;

            case 'confluence_update':
                if (data.meets_threshold) {
                    this.showToast(
                        'High Confluence Alert!',
                        `Score: ${data.total_score}/14 - ${data.themes.join(', ')}`,
                        'success'
                    );
                }
                break;

            case 'theme_update':
                this.showToast(
                    'Theme Updated',
                    `${data.theme_name}: ${data.conviction}/10 (${data.evidence_count} sources)`
                );
                break;

            case 'high_conviction_alert':
                this.showToast(
                    '⚡ High Conviction Alert!',
                    `${data.theme}: ${data.score}/14 (${data.sentiment})`,
                    'success'
                );
                break;
        }
    }

    /**
     * Register event listener
     */
    on(event, callback) {
        if (!this.listeners[event]) {
            this.listeners[event] = [];
        }
        this.listeners[event].push(callback);
    }

    /**
     * Emit event to listeners
     */
    emit(event, ...args) {
        if (this.listeners[event]) {
            this.listeners[event].forEach(callback => {
                try {
                    callback(...args);
                } catch (error) {
                    console.error(`Error in ${event} listener:`, error);
                }
            });
        }
    }

    /**
     * Update connection indicator in UI
     */
    updateConnectionIndicator(connected) {
        const indicator = document.getElementById('ws-connection-indicator');
        if (indicator) {
            indicator.className = connected ? 'update-indicator' : 'update-indicator disconnected';
            indicator.title = connected ? 'Real-time updates active' : 'Disconnected';
        }
    }

    /**
     * Show toast notification
     */
    showToast(title, message, type = 'info') {
        // Remove existing toast
        const existing = document.getElementById('ws-toast');
        if (existing) {
            existing.remove();
        }

        // Create toast
        const toast = document.createElement('div');
        toast.id = 'ws-toast';
        toast.className = 'toast';
        toast.innerHTML = `
            <div style="font-weight: 600; margin-bottom: 4px;">${title}</div>
            <div style="font-size: 13px; color: var(--text-secondary);">${message}</div>
        `;

        // Add type styling
        if (type === 'success') {
            toast.style.borderLeft = '3px solid var(--success)';
        } else if (type === 'error') {
            toast.style.borderLeft = '3px solid var(--danger)';
        }

        document.body.appendChild(toast);

        // Auto-remove after 5 seconds
        setTimeout(() => {
            toast.style.animation = 'slideOut 0.3s ease-out';
            setTimeout(() => toast.remove(), 300);
        }, 5000);
    }
}

// Global WebSocket client
const wsClient = new WebSocketClient();

// Auto-connect on page load
document.addEventListener('DOMContentLoaded', () => {
    wsClient.connect();

    // Add connection indicator to header (if not exists)
    const header = document.querySelector('.header-content');
    if (header && !document.getElementById('ws-connection-indicator')) {
        const indicator = document.createElement('span');
        indicator.id = 'ws-connection-indicator';
        indicator.className = 'update-indicator disconnected';
        indicator.title = 'Disconnected';
        header.appendChild(indicator);
    }
});

// Reconnect on page visibility change
document.addEventListener('visibilitychange', () => {
    if (!document.hidden && wsClient.connectionStatus === 'disconnected') {
        wsClient.connect();
    }
});

File: backend/routes/websocket.py
"""
WebSocket Route for Real-Time Dashboard Updates

Provides real-time push notifications for:
- New content collected
- Analysis completed
- Confluence scores updated
- Theme changes
"""

from fastapi import APIRouter, WebSocket, WebSocketDisconnect
from typing import List
import logging
import json
import asyncio

logger = logging.getLogger(__name__)

router = APIRouter()


class ConnectionManager:
    """Manages active WebSocket connections."""

    def __init__(self):
        self.active_connections: List[WebSocket] = []

    async def connect(self, websocket: WebSocket):
        """Accept and track new WebSocket connection."""
        await websocket.accept()
        self.active_connections.append(websocket)
        logger.info(f"WebSocket connected. Active connections: {len(self.active_connections)}")

    def disconnect(self, websocket: WebSocket):
        """Remove disconnected WebSocket."""
        if websocket in self.active_connections:
            self.active_connections.remove(websocket)
        logger.info(f"WebSocket disconnected. Active connections: {len(self.active_connections)}")

    async def broadcast(self, message: dict):
        """Broadcast message to all connected clients."""
        disconnected = []

        for connection in self.active_connections:
            try:
                await connection.send_json(message)
            except Exception as e:
                logger.error(f"Error broadcasting to client: {e}")
                disconnected.append(connection)

        # Clean up disconnected clients
        for conn in disconnected:
            self.disconnect(conn)


# Global connection manager
manager = ConnectionManager()


@router.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    """
    WebSocket endpoint for real-time updates.

    Messages sent to clients:
    - {"type": "new_analysis", "data": {...}}
    - {"type": "collection_complete", "data": {...}}
    - {"type": "confluence_update", "data": {...}}
    - {"type": "theme_update", "data": {...}}
    """
    await manager.connect(websocket)

    try:
        while True:
            # Keep connection alive by receiving messages
            # Clients can send heartbeat pings
            data = await websocket.receive_text()

            # Echo back for heartbeat
            if data == "ping":
                await websocket.send_text("pong")

    except WebSocketDisconnect:
        manager.disconnect(websocket)
    except Exception as e:
        logger.error(f"WebSocket error: {e}")
        manager.disconnect(websocket)


async def broadcast_new_analysis(analysis_data: dict):
    """
    Broadcast when new content is analyzed.

    Args:
        analysis_data: Analysis result dictionary
    """
    message = {
        "type": "new_analysis",
        "timestamp": analysis_data.get("analyzed_at"),
        "data": {
            "source": analysis_data.get("source"),
            "content_type": analysis_data.get("content_type"),
            "key_themes": analysis_data.get("key_themes", []),
            "sentiment": analysis_data.get("sentiment"),
            "conviction": analysis_data.get("conviction")
        }
    }
    await manager.broadcast(message)


async def broadcast_collection_complete(collection_data: dict):
    """
    Broadcast when data collection completes.

    Args:
        collection_data: Collection result dictionary
    """
    message = {
        "type": "collection_complete",
        "timestamp": collection_data.get("completed_at"),
        "data": {
            "source": collection_data.get("source"),
            "items_collected": collection_data.get("count", 0),
            "status": collection_data.get("status")
        }
    }
    await manager.broadcast(message)


async def broadcast_confluence_update(confluence_data: dict):
    """
    Broadcast when confluence scores are updated.

    Args:
        confluence_data: Confluence score dictionary
    """
    message = {
        "type": "confluence_update",
        "timestamp": confluence_data.get("scored_at"),
        "data": {
            "total_score": confluence_data.get("total_score"),
            "core_score": confluence_data.get("core_score"),
            "meets_threshold": confluence_data.get("meets_threshold"),
            "themes": confluence_data.get("key_themes", [])
        }
    }
    await manager.broadcast(message)


async def broadcast_theme_update(theme_data: dict):
    """
    Broadcast when theme strength changes.

    Args:
        theme_data: Theme update dictionary
    """
    message = {
        "type": "theme_update",
        "timestamp": theme_data.get("updated_at"),
        "data": {
            "theme_name": theme_data.get("name"),
            "conviction": theme_data.get("conviction"),
            "evidence_count": theme_data.get("evidence_count")
        }
    }
    await manager.broadcast(message)


async def broadcast_high_conviction_alert(alert_data: dict):
    """
    Broadcast when new idea crosses high-conviction threshold (>=7/14).

    Args:
        alert_data: Alert data dictionary
    """
    message = {
        "type": "high_conviction_alert",
        "timestamp": alert_data.get("scored_at"),
        "data": {
            "theme": alert_data.get("theme_name"),
            "score": alert_data.get("total_score"),
            "sentiment": alert_data.get("sentiment")
        }
    }
    await manager.broadcast(message)


async def broadcast_synthesis_complete(synthesis_data: dict):
    """
    Broadcast when synthesis generation completes.

    Args:
        synthesis_data: Synthesis result dictionary
    """
    message = {
        "type": "synthesis_complete",
        "timestamp": synthesis_data.get("generated_at"),
        "data": {
            "status": synthesis_data.get("status"),
            "synthesis_id": synthesis_data.get("synthesis_id"),
            "time_window": synthesis_data.get("time_window"),
            "content_count": synthesis_data.get("content_count"),
            "error": synthesis_data.get("error")
        }
    }
    await manager.broadcast(message)
